compiler/qsc_qasm3/src/parser/completion/word_kinds.rs

```diff
@@ -0,0 +1,175 @@
+        const Let = keyword_bit(Keyword::Let);
+        const Measure = keyword_bit(Keyword::Measure);
+        const Mutable = keyword_bit(Keyword::Mutable);
+        const NegCtrl = keyword_bit(Keyword::NegCtrl);
+        const OpenQASM = keyword_bit(Keyword::OpenQASM);
+        const Output = keyword_bit(Keyword::Output);
+        const Pow = keyword_bit(Keyword::Pow);
+        const Pragma = keyword_bit(Keyword::Pragma);
+        const QReg = keyword_bit(Keyword::QReg);
+        const Qubit = keyword_bit(Keyword::Qubit);
+        const Reset = keyword_bit(Keyword::Reset);
+        const True = keyword_bit(Keyword::True);
+        const ReadOnly = keyword_bit(Keyword::ReadOnly);
+        const Return = keyword_bit(Keyword::Return);
+        const Switch = keyword_bit(Keyword::Switch);
+        const Void = keyword_bit(Keyword::Void);
+        const While = keyword_bit(Keyword::While);
+    }
+}
+
+const KEYWORDS_START: u8 = 4;
+const fn keyword_bit(k: Keyword) -> u128 {
+    1 << (k as u8 + KEYWORDS_START)
+}
+
+impl From<Keyword> for WordKinds {
+    fn from(k: Keyword) -> Self {
+        Self::from_bits_truncate(keyword_bit(k))
+    }
+}
+
+impl WordKinds {
+    /// Returns only the name kinds that this prediction set contains.
+    pub fn iter_name_kinds(&self) -> impl Iterator<Item = NameKind> + '_ {
+        self.iter().filter_map(|p| match p {
+            WordKinds::PathExpr => Some(NameKind::Path(PathKind::Expr)),
+            WordKinds::PathSegment => Some(NameKind::PathSegment),
+            _ => None,
+        })
+    }
+
+    /// Returns only the hardcoded identifier kinds that this prediction set contains.
+    pub fn iter_hardcoded_ident_kinds(&self) -> impl Iterator<Item = HardcodedIdentKind> + '_ {
+        self.iter().filter_map(|p| match p {
+            WordKinds::Annotation => Some(HardcodedIdentKind::Annotation),
+            _ => None,
+        })
+    }
+
+    /// Returns only the keywords that this prediction set contains.
+    pub fn iter_keywords(&self) -> impl Iterator<Item = Keyword> + '_ {
+        all::<Keyword>().filter(|k| self.contains((*k).into()))
+    }
+}
+
+/// A hardcoded identifier.
+///
+/// Maps to a subset of values in [`Predictions`], but an enum
+/// for friendly consumption.
+pub enum HardcodedIdentKind {
+    /// An attribute, without the leading `@`.
+    Annotation,
+}
+
+/// A name (see: [`Predictions`])
+///
+/// Maps to a subset of values in [`Predictions`], but an enum
+/// for friendly consumption.
+pub enum NameKind {
+    /// A path.
+    Path(PathKind),
+    /// A path segment that follows a `.`
+    /// A more specific name kind can only be inferred from a recovered AST.
+    PathSegment,
+}
+
+/// A path (see: [`Predictions`])
+///
+/// Maps to a subset of values in [`Predictions`], but an enum
+/// for friendly consumption.
+#[derive(Debug, Clone, Copy)]
+pub enum PathKind {
+    /// A path in an expression. Callables, UDT constructors, local variables.
+    Expr,
+}
```

Comment:
Some of this doesn't make sense to me for QASM, but I think I'm just not going to worry about it until I see their usage from the language service. It all comes down to _how_ this stuff actually gets used in the language service implementation. Once we get there, we may find that these complex types are just not necessary.